sim/pokemon.js

```diff
@@ -1251,13 +1253,13 @@ class Pokemon {
 	}
 
 	/**
-	 * @param {string} abilityName
-	 * @param {Pokemon} [source]
+	 * @param {string | Ability} ability
+	 * @param {?Pokemon} [source]
 	 * @param {boolean} [isFromFormeChange]
 	 */
-	setAbility(abilityName, source, isFromFormeChange) {
+	setAbility(ability, source, isFromFormeChange) {
 		if (!this.hp) return false;
-		let ability = this.battle.getAbility(abilityName);
+		if (typeof ability === 'string') ability = this.battle.getAbility(ability);
```

Comment:
Why is this needed?

Reply:
Not exactly needed, but it's preventing an unnecessary function call if an <code>Ability</code> is passed in.

Reply:
This seems like premature optimization... Surveying for ``.setAbility()`` calls, an ``Ability`` instance is passed almost nowhere (only in Trace for a couple of old gen mods).

Will TS still know that ``ability`` is always an ``Ability`` instance from next line onwards with this change?

Reply:
Yes it will.

Reply:
Yeah, I think the term is "contextual typing" and TypeScript does it and it's amazing and basically the thing about TypeScript I love most. I only wish it were easier to assert a contextual type without providing an explicit guard.